### ISSUE TITLE
Halyard pulling version list fails SSL Cert Path

### ISSUE THREAD

**Issue by Spellchaser** [spinnaker/halyard#688]
I'm trying to get Halyard working on a corp network (read firewalls and cert modifications). I've configured my proxies correctly #558 but I'm pretty sure my corp is messing with SSL certs. 

I've tried installing certs from gcr.io, dl.bintray.com, and storage.googleapis.com into the jks but it hasn't worked. I've seen something about loading BOMs from disk?

What can I do to debug this further?

Seemingly related issues #680 
https://github.com/spinnaker/spinnaker/issues/1773

Error:
```
Problems in Global:
! ERROR Could not load "versions.yml" from config bucket:
  sun.security.validator.ValidatorException: PKIX path building failed:
  sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid
  certification path to requested target.

- Failed to load available Spinnaker versions.

```
Error occurs when running `hal version list ` or anything requiring HalValidator
```
$ hal deploy apply
+ Get current deployment
  Success
- Apply deployment
  Failure
Problems in default.provider.openstack:
- WARNING Provider openstack is enabled, but no accounts have been
  configured.

Problems in Global:
! ERROR Unable to retrieve the Spinnaker bill of materials for
  version "1.2.2": sun.security.validator.ValidatorException: PKIX path building
  failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to
  find valid certification path to requested target

- Failed to deploy Spinnaker.

```

Debug log:
https://gist.github.com/anonymous/71293722aaf9ccdf50969d177923a06c
https://gist.github.com/anonymous/0f0b1920cb679e4422737b0c34f05880

**Comment by pycaster** [comment]
Did you find the solution?

**Comment by fabiogoma** [comment]
Facing the same issues here, I have a internal proxy with inspection certificate, so I had to add the company Root CA to the cacerts, It worked for the private docker registry (artifactory) but didn't work for external access. Troubleshooting I saw the container trying to reach the url https://www.googleapis.com.

**Comment by pycaster** [comment]
@fabiogoma 

How did you load the certs for private docker registry? I am facing a similar issue.

**Comment by fabiogoma** [comment]
Hi @pycaster, sorry for delay.
I'm currently trying to use the containerized version, because all my infrastructure is redhat based.
If you have a internal Root CA (that's my case), you need to get your root certificate, copy it to the container and the execute the keytool command to import it to your keystore.

Example:
\# keytool -import -trustcacerts -alias rootAliasName -file yourCertificatePemFormat.cer -keystore /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts -noprompt -storepass changeit

**Comment by fabiogoma** [comment]
I have finally nailed this issue, at least for my case. The problem was the container DNS servers, I changed them both to my internal DNS servers (/etc/resolv.conf), after execute the command again it worked.

**Comment by MoShitrit** [comment]
+1 on that issue.
Here's my scenario:

I am currently running halyard as a container. We are using a proxy server to access the internet in our environment, which also means that it's altering with SSL traffic (re-encrypts traffic using our internal CA). 
Steps I performed on the container:
1. Modified `/opt/halyard/bin/halyard` file, `DEFAULT_JVM_OPTS`. Added: `-Dhttp.proxyHost=proxy.domain.com" "-Dhttp.proxyPort=<port>" "-Dhttps.proxyHost=proxy.domain.com" "-Dhttps.proxyPort=<port>"'.`
2. Exported our root CA in pem format and converted to der by running:
   `openssl x509 -in ca.pem -inform pem -out ca.der -outform der`
3. `keytool -importcert -alias ca -keystore /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts -storepass **** -file ca.der`
Confirmed that the CA indeed exists on the cacerts file afterwards.
4. Restarted the container, still getting the same error:
```
root@dbd9717ba546:/workdir# hal version list
+ Get current deployment
  Success
+ Get Spinnaker version
  Success
- Get released versions
  Failure
Problems in Global:
! ERROR Could not load "versions.yml" from config bucket:
  sun.security.validator.ValidatorException: PKIX path building failed:
  sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid
  certification path to requested target.

- Failed to load available Spinnaker versions.
```
Based on what @fabiogoma suggested, also tried to modify my `keytool` command to `keytool -import -trustcacerts -alias rootAliasName -file yourCertificatePemFormat.cer -keystore /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts -noprompt -storepass changeit`.

Still failing..

Anyone has any suggestions?
